test/functional/services/filter_bar.ts

```diff
@@ -14,7 +14,7 @@ export function FilterBarProvider({ getService, getPageObjects }: FtrProviderCon
   const testSubjects = getService('testSubjects');
   const PageObjects = getPageObjects(['common', 'header']);
 
-  class FilterBar {
+  return {
```

Comment:
@spalger the problem exists because TS cannot infer a result type, we should either declare the result type explicitly
```typescript
export interface ISomeType {
  someMethod(): Promise<void>
}
export function SomeProvider (deps: FtrProviderContext): ISomeType {
 class SomeClass {}
 return new SomeClass()
}
```
Or allow TS to infer the type. We've got at least two options here:
- use POJO as you refactored in the current PR
- export classes

All the options have different trade-offs.
**Declaring a return type**: manual work, easy to forget to update the class, `Go to the declaration` in IDE navigates to an interface instead of implementation. 
**Return POJO**: we lose namespace information, which worsens DX - adds visual noise in tooltips, a developer cannot navigate to a service class to inspect it, but have to navigate to each method instead. 
![2021-05-04_10-32-30](https://user-images.githubusercontent.com/3198181/116978663-4ec47f00-acc4-11eb-863f-2a71cf2df79c.png)
**Export class explicitly**: the recommended way to write code in TS. But it requires more work than the other alternatives: we will have to declare type deps explicitly and refactor class declaration not to use closures for deps. 
before:
```typescript
const testSubjects = getService('testSubjects');
return (new MyClass{
  doSomething() {
     return testSubjects.exists(...);
  }
})()
```

after:
```typescript
import type TypeSubjects from './path-to-test-subjects';

export new MyClass {
  constructor(testSubjects: TypeSubjects){}
  doSomething() {
     return this.testSubjects.exists(...);
  }
}

export function provider() {
  const testSubjects = getService('testSubjects');
  return new MyClass(testSubjects);
}
```

It seems the implementation you picked in the current PR requires less work, even though I'd prefer an approach with `export class` for explicitness.

Reply:
Yeah, how do you feel about this: https://github.com/elastic/kibana/blob/de17ce198808d4cee6d0f20ae46c999a6d61b485/test/functional/page_objects/console_page.ts#L13-L23

In this example I'm keeping all the types locked up in the `FtrProviderContext` rather than importing all the service types directly.

Reply:
I think it's probably best to go with the "Export class explicitly" approach, how do you feel about this: https://github.com/spalger/kibana/blob/implement/root-test-ts-project/test/functional/page_objects/context_page.ts#L17-L44

The base class `FtrService` is setup [here](https://github.com/spalger/kibana/blob/implement/root-test-ts-project/test/functional/ftr_provider_context.ts#L15) where it is configured with all the services available for this FTR config, and it assigns the `FtrProviderContext` to `this.ctx` so we can use it to capture services/pageObjects as instance properties of the service classes. Additionally, the service collection was updated in [the latest commit](https://github.com/elastic/kibana/pull/99148/commits/c80ffad65ff8984e97a741063760fa19e2da001a#diff-67ccd8a50a00bd8c41ff51e77302cbd069c73cd8bd59be05ee5131b0237e06d9R69-R72) to instantiate those classes with the right context argument so we can just migrate away from providers all together.

Reply:
The primary thing I don't like about this approach is the amount of changes it requires, though they're almost completely mindless updates that I can crank out pretty quickly. The diff doesn't even look half bad if you disable whitespace: 
![image](https://user-images.githubusercontent.com/1329312/117093476-ee5e2d80-ad15-11eb-984f-bc4d31aab973.png)